spec/integration/github_omniauth_spec.rb

```diff
@@ -0,0 +1,232 @@
+          subscriptions_url: "https://api.github.com/users/octocat/subscriptions",
+          organizations_url: "https://api.github.com/users/octocat/orgs",
+          repos_url: "https://api.github.com/users/octocat/repos",
+          events_url: "https://api.github.com/users/octocat/events{/privacy}",
+          received_events_url: "https://api.github.com/users/octocat/received_events",
+          type: "User",
+          site_admin: false,
+          name: "monalisa octocat",
+          company: "GitHub",
+          blog: "https://github.com/blog",
+          location: "San Francisco",
+          email: "octocat@github.com",
+          hireable: false,
+          bio: "There once was...",
+          twitter_username: "monatheoctocat",
+          public_repos: 2,
+          public_gists: 1,
+          followers: 20,
+          following: 0,
+          created_at: "2008-01-14T04:33:35Z",
+          updated_at: "2008-01-14T04:33:35Z",
+          private_gists: 81,
+          total_private_repos: 100,
+          owned_private_repos: 100,
+          disk_usage: 10000,
+          collaborators: 8,
+          two_factor_authentication: true,
+          plan: {
+            name: "Medium",
+            space: 400,
+            private_repos: 20,
+            collaborators: 0
+          }
+        ),
+        headers: {
+          "Content-Type" => "application/json"
+        }
+      )
+  end
+
+  it "doesn't sign in anyone if none of the emails from github are verified" do
+    post "/auth/github"
+
+    setup_github_emails_stub(
+      [
+        {
+          email: user1.email,
+          primary: true,
+          verified: false,
+          visibility: "private"
+        },
+        {
+          email: user2.email,
+          primary: false,
+          verified: false,
+          visibility: "private"
+        }
+      ]
+    )
+
+    post "/auth/github/callback", params: {
+      state: session["omniauth.state"],
+      code: temp_code
+    }
+    expect(response.status).to eq(302)
+    expect(response.location).to eq("http://test.localhost/")
+    expect(cookies[:_t]).to be_blank
+  end
+
+  it "matches a non-primary email if it's verified and the primary email isn't" do
+    post "/auth/github"
+
+    setup_github_emails_stub(
+      [
+        {
+          email: user1.email,
+          primary: true,
+          verified: false,
+          visibility: "private"
+        },
+        {
+          email: user2.email,
+          primary: false,
+          verified: true,
+          visibility: "private"
+        }
+      ]
+    )
+
+    post "/auth/github/callback", params: {
+      state: session["omniauth.state"],
+      code: temp_code
+    }
+    expect(response.status).to eq(302)
+    expect(response.location).to eq("http://test.localhost/")
+    expect(
+      UserAuthToken.lookup(request.cookie_jar.encrypted[:_t]["token"]).user.id
+    ).to eq(user2.id)
```

Comment:
I'm not quite familiar with the code here but what does creating the `UserAuthToken` do?

Reply:
When a user is logged in, a `UserAuthToken` record is created which contains the user's auth cookie (hashed of course) and some information about the user's session. I'm asserting here that the auth cookie that is added by the request belongs to the correct user. However, I've now switched to use `session[:current_user_id]` for my assertion which is simpler/shorter.